src/cpp/rtps/writer/StatefulWriter.cpp

```diff
@@ -274,9 +305,26 @@ void StatefulWriter::unsent_change_added_to_history(
                         {
                             RTPSMessageGroup group(mp_RTPSParticipant, this, it->message_sender(),
                                     max_blocking_time);
-                            if (!group.add_data(*change, it->expects_inline_qos()))
+
+                            if (change->getFragmentCount() > 0)
                             {
-                                logError(RTPS_WRITER, "Error sending change " << change->sequenceNumber);
+                                ChangeForReader_t change_for_reader(change);
+                                change_for_reader.getUnsentFragments().for_each([&](FragmentNumber_t fragment_number)
+                                {
+                                    if (!group.add_data_frag(*change, fragment_number,
+                                            it->expects_inline_qos()))
+                                    {
+                                        logError(RTPS_WRITER, "Error sending fragment (" << change->sequenceNumber <<
+                                            ", " << fragment_number << ")");
+                                    }
```

Comment:
Should we add an `else` here and mark the fragment as sent, the same way it is done on line 255?